File: apps/reading/src/api/writeModeTwinCollectiveAnalysisCompose.ts
/**
 * Write-mode twin draft + collective analysis pack (pure).
 *
 * Operator vision: after subagent chases complete, merge findings into written
 * analysis; also fold twin note-taker substrate into a provisional write draft
 * before full author commit — HTML-native writing surface.
 *
 * draft_written always false.
 * analysis_written always false.
 * merge_executed always false.
 * store_mutated always false.
 * live_dispatched always false.
 */

import {
  composeWriteModeTwinDraftMerge,
  type TwinWriteSlice,
  type WriteModeTwinDraftMergeCompose,
} from "./writeModeTwinDraftMergeCompose";
import {
  composeChaseCompletionCollectiveAnalysis,
  type ChaseCompletionCollectiveAnalysisCompose,
  type CompletedChaseSlot,
} from "./chaseCompletionCollectiveAnalysisCompose";
import type { AnalysisMergeKind } from "./collectiveDeepResearchMerge";

export interface WriteModeTwinCollectiveAnalysisInput {
  session_id: string;
  draft_id: string;
  parent_asset_id: string;
  /** Twin slices for write draft (caller-supplied). */
  twin_slices: TwinWriteSlice[];
  base_draft_html?: string | null;
  /** Completed chase slots for collective analysis. */
  chase_slots: CompletedChaseSlot[];
  analysis_kind: AnalysisMergeKind;
  extra_findings?: string[] | null;
  operator_ack: boolean;
  /**
   * When true (default), require both draft_ready and analysis_ready for pack.
   * When false, pack_ready if either path ready.
   */
  require_both?: boolean;
}

export interface WriteModeTwinCollectiveAnalysisCompose {
  session_id: string;
  draft_id: string;
  parent_asset_id: string;
  twin_draft: WriteModeTwinDraftMergeCompose;
  collective_analysis: ChaseCompletionCollectiveAnalysisCompose;
  /**
   * True when twin_draft.draft_ready and collective_analysis.analysis_ready
   * (or either when require_both=false) and operator_ack.
   */
  pack_ready: boolean;
  draft_written: false;
  analysis_written: false;
  merge_executed: false;
  store_mutated: false;
  live_dispatched: false;
  notes: string[];
  authority: "write_mode_twin_collective_analysis_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose twin→write draft + chase-completion collective analysis.
 * Never writes draft/analysis assets; never merges published write.
 */
export function composeWriteModeTwinCollectiveAnalysis(
  input: WriteModeTwinCollectiveAnalysisInput,
): WriteModeTwinCollectiveAnalysisCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const draft_id = requireNonEmpty(input.draft_id, "draft_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "draft_written=false — provisional write draft not persisted",
    "analysis_written=false — collective analysis intent only",
    "merge_executed=false — published write not mutated",
    "store_mutated=false · live_dispatched=false",
  ];

  const twin_draft = composeWriteModeTwinDraftMerge({
    draft_id,
    base_draft_html: input.base_draft_html,
    slices: input.twin_slices,
    operator_ack: input.operator_ack,
  });
  notes.push(...twin_draft.notes.map((n) => `[twin_draft] ${n}`));

  // Ensure all chase slots share parent_asset_id
  const slots = input.chase_slots;
  if (!Array.isArray(slots) || slots.length < 2) {
    throw new Error("chase_slots must be an array with at least 2 slots");
  }
  for (let i = 0; i < slots.length; i++) {
    const s = slots[i];
    if (!s || typeof s !== "object") {
      throw new Error(`chase_slots[${i}] must be an object`);
    }
    if (s.parent_asset_id.trim() !== parent_asset_id) {
      throw new Error(
        "all chase_slots must share input.parent_asset_id",
      );
    }
  }

  const collective_analysis = composeChaseCompletionCollectiveAnalysis({
    session_id,
    parent_asset_id,
    slots,
    kind: input.analysis_kind,
    operator_ack: input.operator_ack,
    extra_findings: input.extra_findings,
  });
  notes.push(
    ...collective_analysis.notes.map((n) => `[collective] ${n}`),
  );

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      twin_draft.draft_ready === true &&
      collective_analysis.analysis_ready === true &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      (twin_draft.draft_ready === true ||
        collective_analysis.analysis_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — twin write draft + collective analysis intent ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — twin draft, collective analysis, or operator_ack gate open",
    );
  }

  if (
    twin_draft.draft_written !== false ||
    twin_draft.merge_executed !== false ||
    twin_draft.store_mutated !== false ||
    collective_analysis.analysis_written !== false ||
    collective_analysis.live_dispatched !== false ||
    collective_analysis.pack_dispatched !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("merge_executed=false");
  notes.push("store_mutated=false");
  notes.push("live_dispatched=false");

  return {
    session_id,
    draft_id,
    parent_asset_id,
    twin_draft,
    collective_analysis,
    pack_ready,
    draft_written: false,
    analysis_written: false,
    merge_executed: false,
    store_mutated: false,
    live_dispatched: false,
    notes,
    authority: "write_mode_twin_collective_analysis_compose_advisory",
  };
}

export function formatWriteModeTwinCollectiveAnalysisSummary(
  c: WriteModeTwinCollectiveAnalysisCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · draft_ready=${c.twin_draft.draft_ready} · ` +
    `analysis_ready=${c.collective_analysis.analysis_ready} · ` +
    `kind=${c.collective_analysis.analysis.kind} · ` +
    `draft_written=false · analysis_written=false · merge_executed=false`
  );
}

File: substrate/write_mode_twin_collective_analysis_compose.py
"""Write-mode twin draft + collective analysis pack (pure).

draft_written, analysis_written, merge_executed, store_mutated,
live_dispatched always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.chase_completion_collective_analysis_compose import (
    ChaseCompletionCollectiveAnalysisCompose,
    ChaseCompletionCollectiveAnalysisComposeError,
    compose_chase_completion_collective_analysis,
)
from substrate.write_mode_twin_draft_merge_compose import (
    WriteModeTwinDraftMergeCompose,
    WriteModeTwinDraftMergeComposeError,
    compose_write_mode_twin_draft_merge,
)


class WriteModeTwinCollectiveAnalysisComposeError(ValueError):
    """Fail-closed validation for write-mode twin + collective analysis."""


@dataclass(frozen=True)
class WriteModeTwinCollectiveAnalysisCompose:
    session_id: str
    draft_id: str
    parent_asset_id: str
    twin_draft: WriteModeTwinDraftMergeCompose
    collective_analysis: ChaseCompletionCollectiveAnalysisCompose
    pack_ready: bool
    draft_written: bool
    analysis_written: bool
    merge_executed: bool
    store_mutated: bool
    live_dispatched: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "draft_id": self.draft_id,
            "parent_asset_id": self.parent_asset_id,
            "twin_draft": self.twin_draft.to_dict(),
            "collective_analysis": self.collective_analysis.to_dict(),
            "pack_ready": self.pack_ready,
            "draft_written": False,
            "analysis_written": False,
            "merge_executed": False,
            "store_mutated": False,
            "live_dispatched": False,
            "notes": list(self.notes),
            "authority": (
                "write_mode_twin_collective_analysis_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WriteModeTwinCollectiveAnalysisComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_write_mode_twin_collective_analysis(
    *,
    session_id: object,
    draft_id: object,
    parent_asset_id: object,
    twin_slices: object,
    chase_slots: object,
    analysis_kind: object,
    operator_ack: object,
    base_draft_html: object | None = None,
    extra_findings: object | None = None,
    require_both: object | None = None,
) -> WriteModeTwinCollectiveAnalysisCompose:
    """Twin write draft + collective analysis. Never writes/merges assets."""
    if not isinstance(operator_ack, bool):
        raise WriteModeTwinCollectiveAnalysisComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    draft = _require_nonempty(draft_id, field="draft_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    both = True if require_both is None else require_both
    if not isinstance(both, bool):
        raise WriteModeTwinCollectiveAnalysisComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "draft_written=false — provisional write draft not persisted",
        "analysis_written=false — collective analysis intent only",
        "merge_executed=false — published write not mutated",
        "store_mutated=false · live_dispatched=false",
    ]

    try:
        twin_draft = compose_write_mode_twin_draft_merge(
            draft_id=draft,
            slices=twin_slices,
            operator_ack=operator_ack,
            base_draft_html=base_draft_html,
        )
    except WriteModeTwinDraftMergeComposeError as e:
        raise WriteModeTwinCollectiveAnalysisComposeError(str(e)) from e
    notes.extend(f"[twin_draft] {n}" for n in twin_draft.notes)

    if not isinstance(chase_slots, list) or len(chase_slots) < 2:
        raise WriteModeTwinCollectiveAnalysisComposeError(
            "chase_slots must be an array with at least 2 slots"
        )
    for i, s in enumerate(chase_slots):
        if not isinstance(s, dict):
            raise WriteModeTwinCollectiveAnalysisComposeError(
                f"chase_slots[{i}] must be an object"
            )
        p = s.get("parent_asset_id")
        if not isinstance(p, str) or p.strip() != parent:
            raise WriteModeTwinCollectiveAnalysisComposeError(
                "all chase_slots must share input.parent_asset_id"
            )

    try:
        collective_analysis = compose_chase_completion_collective_analysis(
            session_id=session,
            parent_asset_id=parent,
            slots=chase_slots,
            kind=analysis_kind,
            operator_ack=operator_ack,
            extra_findings=extra_findings,
        )
    except ChaseCompletionCollectiveAnalysisComposeError as e:
        raise WriteModeTwinCollectiveAnalysisComposeError(str(e)) from e
    notes.extend(f"[collective] {n}" for n in collective_analysis.notes)

    if both:
        pack_ready = (
            twin_draft.draft_ready is True
            and collective_analysis.analysis_ready is True
            and operator_ack is True
        )
    else:
        pack_ready = operator_ack is True and (
            twin_draft.draft_ready is True
            or collective_analysis.analysis_ready is True
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — twin write draft + collective analysis intent "
            "ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — twin draft, collective analysis, or "
            "operator_ack gate open"
        )

    if (
        twin_draft.draft_written is not False
        or twin_draft.merge_executed is not False
        or twin_draft.store_mutated is not False
        or collective_analysis.analysis_written is not False
        or collective_analysis.live_dispatched is not False
        or collective_analysis.pack_dispatched is not False
    ):
        raise WriteModeTwinCollectiveAnalysisComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "draft_written=false",
            "analysis_written=false",
            "merge_executed=false",
            "store_mutated=false",
            "live_dispatched=false",
        )
    )

    return WriteModeTwinCollectiveAnalysisCompose(
        session_id=session,
        draft_id=draft,
        parent_asset_id=parent,
        twin_draft=twin_draft,
        collective_analysis=collective_analysis,
        pack_ready=pack_ready,
        draft_written=False,
        analysis_written=False,
        merge_executed=False,
        store_mutated=False,
        live_dispatched=False,
        notes=tuple(notes),
        authority="write_mode_twin_collective_analysis_compose_advisory",
    )


def format_write_mode_twin_collective_analysis_summary(
    c: WriteModeTwinCollectiveAnalysisCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · draft_ready={c.twin_draft.draft_ready} · "
        f"analysis_ready={c.collective_analysis.analysis_ready} · "
        f"kind={c.collective_analysis.analysis.kind} · "
        f"draft_written=false · analysis_written=false · merge_executed=false"
    )


__all__ = [
    "WriteModeTwinCollectiveAnalysisCompose",
    "WriteModeTwinCollectiveAnalysisComposeError",
    "compose_write_mode_twin_collective_analysis",
    "format_write_mode_twin_collective_analysis_summary",
]
